Architecture support: build failures on architectures other than x86_64

While [building for Fedora](https://koji.fedoraproject.org/koji/taskinfo?taskID=39482950), I've noticed that arbor fails to build on all architectures other than x86_64. These other architectures we currently build in Fedora for are: armv7hl, i686, aarch64, ppc64le, s390x.

The crash log says:

```
c++: warning: '-mcpu=' is deprecated; use '-mtune=' or '-march=' instead
In file included from /builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/arbexcept.hpp:6,
                 from /builddir/build/BUILD/arbor-0.2.2/arbor/arbexcept.cpp:4:
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp: In member function 'std::size_t std::hash<arb::cell_member_type>::operator()(const arb::cell_member_type&) const':
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp:102:46: error: static assertion failed: invalid size assumptions for hash of cell_member_type
  102 |             static_assert(sizeof(std::size_t)>sizeof(cell_gid_type), "invalid size assumptions for hash of cell_member_type");
      |                           ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp:104:76: warning: left shift count >= width of type [-Wshift-count-overflow]
  104 |             std::size_t k = ((std::size_t)m.gid << (8*sizeof(cell_gid_type))) + m.index;
      |                                                                            ^
make[2]: *** [arbor/CMakeFiles/arbor.dir/build.make:66: arbor/CMakeFiles/arbor.dir/arbexcept.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Entering directory '/builddir/build/BUILD/arbor-0.2.2/build-serial'
[ 32%] Building CXX object arbor/CMakeFiles/arbor.dir/backends/multicore/stimulus.cpp.o
cd /builddir/build/BUILD/arbor-0.2.2/build-serial/arbor && /usr/bin/c++   -I/builddir/build/BUILD/arbor-0.2.2/arbor -I/builddir/build/BUILD/arbor-0.2.2/arbor/include -I/builddir/build/BUILD/arbor-0.2.2/build-serial/arbor/include  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m32 -march=i686 -mtune=generic -msse2 -mfpmath=sse -mstackrealign -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -DNDEBUG -fPIC   -g -Wall -Wno-maybe-uninitialized -mcpu=native -std=gnu++14 -o CMakeFiles/arbor.dir/backends/multicore/stimulus.cpp.o -c /builddir/build/BUILD/arbor-0.2.2/arbor/backends/multicore/stimulus.cpp
make[2]: Leaving directory '/builddir/build/BUILD/arbor-0.2.2/build-serial'
c++: warning: '-mcpu=' is deprecated; use '-mtune=' or '-march=' instead
In file included from /builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/fvm_types.hpp:3,
                 from /builddir/build/BUILD/arbor-0.2.2/arbor/backends/multicore/stimulus.cpp:3:
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp: In member function 'std::size_t std::hash<arb::cell_member_type>::operator()(const arb::cell_member_type&) const':
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp:102:46: error: static assertion failed: invalid size assumptions for hash of cell_member_type
  102 |             static_assert(sizeof(std::size_t)>sizeof(cell_gid_type), "invalid size assumptions for hash of cell_member_type");
      |                           ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp:104:76: warning: left shift count >= width of type [-Wshift-count-overflow]
  104 |             std::size_t k = ((std::size_t)m.gid << (8*sizeof(cell_gid_type))) + m.index;
      |                                                                            ^
make[2]: *** [arbor/CMakeFiles/arbor.dir/build.make:118: arbor/CMakeFiles/arbor.dir/backends/multicore/stimulus.cpp.o] Error 1
make[2]: Entering directory '/builddir/build/BUILD/arbor-0.2.2/build-serial'
[ 32%] Building CXX object arbor/CMakeFiles/arbor.dir/backends/multicore/shared_state.cpp.o
cd /builddir/build/BUILD/arbor-0.2.2/build-serial/arbor && /usr/bin/c++   -I/builddir/build/BUILD/arbor-0.2.2/arbor -I/builddir/build/BUILD/arbor-0.2.2/arbor/include -I/builddir/build/BUILD/arbor-0.2.2/build-serial/arbor/include  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m32 -march=i686 -mtune=generic -msse2 -mfpmath=sse -mstackrealign -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -DNDEBUG -fPIC   -g -Wall -Wno-maybe-uninitialized -mcpu=native -std=gnu++14 -o CMakeFiles/arbor.dir/backends/multicore/shared_state.cpp.o -c /builddir/build/BUILD/arbor-0.2.2/arbor/backends/multicore/shared_state.cpp
make[2]: Leaving directory '/builddir/build/BUILD/arbor-0.2.2/build-serial'
c++: warning: '-mcpu=' is deprecated; use '-mtune=' or '-march=' instead
In file included from /builddir/build/BUILD/arbor-0.2.2/arbor/backends/multicore/shared_state.cpp:11:
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp: In member function 'std::size_t std::hash<arb::cell_member_type>::operator()(const arb::cell_member_type&) const':
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp:102:46: error: static assertion failed: invalid size assumptions for hash of cell_member_type
  102 |             static_assert(sizeof(std::size_t)>sizeof(cell_gid_type), "invalid size assumptions for hash of cell_member_type");
      |                           ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp:104:76: warning: left shift count >= width of type [-Wshift-count-overflow]
  104 |             std::size_t k = ((std::size_t)m.gid << (8*sizeof(cell_gid_type))) + m.index;
      |                                                                            ^
make[2]: *** [arbor/CMakeFiles/arbor.dir/build.make:105: arbor/CMakeFiles/arbor.dir/backends/multicore/shared_state.cpp.o] Error 1
make[2]: Entering directory '/builddir/build/BUILD/arbor-0.2.2/build-serial'
[ 30%] Building CXX object arbor/CMakeFiles/arbor.dir/backends/multicore/mechanism.cpp.o
cd /builddir/build/BUILD/arbor-0.2.2/build-serial/arbor && /usr/bin/c++   -I/builddir/build/BUILD/arbor-0.2.2/arbor -I/builddir/build/BUILD/arbor-0.2.2/arbor/include -I/builddir/build/BUILD/arbor-0.2.2/build-serial/arbor/include  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m32 -march=i686 -mtune=generic -msse2 -mfpmath=sse -mstackrealign -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -DNDEBUG -fPIC   -g -Wall -Wno-maybe-uninitialized -mcpu=native -std=gnu++14 -o CMakeFiles/arbor.dir/backends/multicore/mechanism.cpp.o -c /builddir/build/BUILD/arbor-0.2.2/arbor/backends/multicore/mechanism.cpp
make[2]: Leaving directory '/builddir/build/BUILD/arbor-0.2.2/build-serial'
c++: warning: '-mcpu=' is deprecated; use '-mtune=' or '-march=' instead
In file included from /builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/fvm_types.hpp:3,
                 from /builddir/build/BUILD/arbor-0.2.2/arbor/backends/multicore/mechanism.cpp:8:
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp: In member function 'std::size_t std::hash<arb::cell_member_type>::operator()(const arb::cell_member_type&) const':
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp:102:46: error: static assertion failed: invalid size assumptions for hash of cell_member_type
  102 |             static_assert(sizeof(std::size_t)>sizeof(cell_gid_type), "invalid size assumptions for hash of cell_member_type");
      |                           ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp:104:76: warning: left shift count >= width of type [-Wshift-count-overflow]
  104 |             std::size_t k = ((std::size_t)m.gid << (8*sizeof(cell_gid_type))) + m.index;
      |                                                                            ^
make[2]: *** [arbor/CMakeFiles/arbor.dir/build.make:92: arbor/CMakeFiles/arbor.dir/backends/multicore/mechanism.cpp.o] Error 1
make[2]: Entering directory '/builddir/build/BUILD/arbor-0.2.2/build-serial'
[ 35%] Building CXX object arbor/CMakeFiles/arbor.dir/communication/dry_run_context.cpp.o
cd /builddir/build/BUILD/arbor-0.2.2/build-serial/arbor && /usr/bin/c++   -I/builddir/build/BUILD/arbor-0.2.2/arbor -I/builddir/build/BUILD/arbor-0.2.2/arbor/include -I/builddir/build/BUILD/arbor-0.2.2/build-serial/arbor/include  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m32 -march=i686 -mtune=generic -msse2 -mfpmath=sse -mstackrealign -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -DNDEBUG -fPIC   -g -Wall -Wno-maybe-uninitialized -mcpu=native -std=gnu++14 -o CMakeFiles/arbor.dir/communication/dry_run_context.cpp.o -c /builddir/build/BUILD/arbor-0.2.2/arbor/communication/dry_run_context.cpp
make[2]: Leaving directory '/builddir/build/BUILD/arbor-0.2.2/build-serial'
c++: warning: '-mcpu=' is deprecated; use '-mtune=' or '-march=' instead
In file included from /builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/spike.hpp:6,
                 from /builddir/build/BUILD/arbor-0.2.2/arbor/communication/dry_run_context.cpp:5:
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp: In member function 'std::size_t std::hash<arb::cell_member_type>::operator()(const arb::cell_member_type&) const':
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp:102:46: error: static assertion failed: invalid size assumptions for hash of cell_member_type
  102 |             static_assert(sizeof(std::size_t)>sizeof(cell_gid_type), "invalid size assumptions for hash of cell_member_type");
      |                           ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp:104:76: warning: left shift count >= width of type [-Wshift-count-overflow]
  104 |             std::size_t k = ((std::size_t)m.gid << (8*sizeof(cell_gid_type))) + m.index;
      |                                                                            ^
make[2]: *** [arbor/CMakeFiles/arbor.dir/build.make:144: arbor/CMakeFiles/arbor.dir/communication/dry_run_context.cpp.o] Error 1
make[2]: Entering directory '/builddir/build/BUILD/arbor-0.2.2/build-serial'
[ 32%] Building CXX object arbor/CMakeFiles/arbor.dir/communication/communicator.cpp.o
cd /builddir/build/BUILD/arbor-0.2.2/build-serial/arbor && /usr/bin/c++   -I/builddir/build/BUILD/arbor-0.2.2/arbor -I/builddir/build/BUILD/arbor-0.2.2/arbor/include -I/builddir/build/BUILD/arbor-0.2.2/build-serial/arbor/include  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m32 -march=i686 -mtune=generic -msse2 -mfpmath=sse -mstackrealign -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -DNDEBUG -fPIC   -g -Wall -Wno-maybe-uninitialized -mcpu=native -std=gnu++14 -o CMakeFiles/arbor.dir/communication/communicator.cpp.o -c /builddir/build/BUILD/arbor-0.2.2/arbor/communication/communicator.cpp
c++: warning: '-mcpu=' is deprecated; use '-mtune=' or '-march=' instead
In file included from /builddir/build/BUILD/arbor-0.2.2/arbor/communication/communicator.cpp:5:
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp: In member function 'std::size_t std::hash<arb::cell_member_type>::operator()(const arb::cell_member_type&) const':
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp:102:46: error: static assertion failed: invalid size assumptions for hash of cell_member_type
  102 |             static_assert(sizeof(std::size_t)>sizeof(cell_gid_type), "invalid size assumptions for hash of cell_member_type");
      |                           ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/arbor-0.2.2/arbor/include/arbor/common_types.hpp:104:76: warning: left shift count >= width of type [-Wshift-count-overflow]
  104 |             std::size_t k = ((std::size_t)m.gid << (8*sizeof(cell_gid_type))) + m.index;
      |                                                                            ^
make[2]: *** [arbor/CMakeFiles/arbor.dir/build.make:131: arbor/CMakeFiles/arbor.dir/communication/communicator.cpp.o] Error 1
make[2]: Leaving directory '/builddir/build/BUILD/arbor-0.2.2/build-serial'
make[1]: *** [CMakeFiles/Makefile2:637: arbor/CMakeFiles/arbor.dir/all] Error 2
make: *** [Makefile:133: all] Error 2
```

It seems to be the same error on all these failing arches. The complete build logs can be found here: https://koji.fedoraproject.org/koji/taskinfo?taskID=39482950. 

Does arbor support these arches? (We can exclude them from the Fedora build if it does not.) I use `-DARB_ARCH=native` in the builds, if that makes a difference.

Thanks,

--- a/arbor/include/arbor/common_types.hpp
+++ b/arbor/include/arbor/common_types.hpp
@@ -99,10 +99,24 @@ namespace std {
     template <> struct hash<arb::cell_member_type> {
         std::size_t operator()(const arb::cell_member_type& m) const {
             using namespace arb;
-            static_assert(sizeof(std::size_t)>sizeof(cell_gid_type), "invalid size assumptions for hash of cell_member_type");
+            if (sizeof(std::size_t)>sizeof(cell_gid_type)) {
+                constexpr unsigned shift = 8*sizeof(cell_gid_type);
 
-            std::size_t k = ((std::size_t)m.gid << (8*sizeof(cell_gid_type))) + m.index;
-            return std::hash<std::size_t>{}(k);
+                std::size_t k = m.gid;
+                k <<= (shift/2); // dodge gcc shift warning when other branch taken
+                k <<= (shift/2);
+                k += m.index;
+                return std::hash<std::size_t>{}(k);
+            }
+            else {
+                constexpr std::size_t prime1 = 93481;
+                constexpr std::size_t prime2 = 54517;
+
+                std::size_t k = prime1;
+                k = k*prime2 + m.gid;
+                k = k*prime2 + m.index;
+                return k;
+            }
         }
     };
 }

--- a/cmake/CompilerOptions.cmake
+++ b/cmake/CompilerOptions.cmake
@@ -91,7 +91,7 @@ function(set_arch_target optvar arch)
 
         # Use -mcpu for all supported targets _except_ for x86, where it should be -march.
 
-        if(target_model MATCHES "x86" OR target_model MATCHES "amd64" OR target_model MATCHES "aarch64")
+        if(target_model MATCHES "x86|i[3456]86" OR target_model MATCHES "amd64" OR target_model MATCHES "aarch64")
             set(arch_opt "-march=${arch}")
         else()
             set(arch_opt "-mcpu=${arch}")

--- a/test/unit/test_algorithms.cpp
+++ b/test/unit/test_algorithms.cpp
@@ -663,7 +663,7 @@ TEST(algorithms, binary_find)
         auto ita = binary_find(a, 1);
         auto found = ita!=std::end(a);
         EXPECT_TRUE(found);
-        EXPECT_EQ(std::distance(std::begin(a), ita), 0u);
+        EXPECT_EQ(std::distance(std::begin(a), ita), 0);
         if (found) {
             EXPECT_EQ(*ita, 1);
         }
@@ -672,7 +672,7 @@ TEST(algorithms, binary_find)
         auto itv = binary_find(v, 1);
         found = itv!=std::end(v);
         EXPECT_TRUE(found);
-        EXPECT_EQ(std::distance(std::begin(v), itv), 0u);
+        EXPECT_EQ(std::distance(std::begin(v), itv), 0);
         if (found) {
             EXPECT_EQ(*itv, 1);
         }
@@ -684,7 +684,7 @@ TEST(algorithms, binary_find)
         auto ita = binary_find(a, 15);
         auto found = ita!=std::end(a);
         EXPECT_TRUE(found);
-        EXPECT_EQ(std::distance(std::begin(a), ita), 2u);
+        EXPECT_EQ(std::distance(std::begin(a), ita), 2);
         if (found) {
             EXPECT_EQ(*ita, 15);
         }
@@ -693,7 +693,7 @@ TEST(algorithms, binary_find)
         auto itv = binary_find(v, 15);
         found = itv!=std::end(v);
         EXPECT_TRUE(found);
-        EXPECT_EQ(std::distance(std::begin(v), itv), 2u);
+        EXPECT_EQ(std::distance(std::begin(v), itv), 2);
         if (found) {
             EXPECT_EQ(*itv, 15);
         }
@@ -705,7 +705,7 @@ TEST(algorithms, binary_find)
         auto ita = binary_find(a, 10);
         auto found = ita!=std::end(a);
         EXPECT_TRUE(found);
-        EXPECT_EQ(std::distance(std::begin(a), ita), 1u);
+        EXPECT_EQ(std::distance(std::begin(a), ita), 1);
         if (found) {
             EXPECT_EQ(*ita, 10);
         }
@@ -714,7 +714,7 @@ TEST(algorithms, binary_find)
         auto itv = binary_find(v, 10);
         found = itv!=std::end(v);
         EXPECT_TRUE(found);
-        EXPECT_EQ(std::distance(std::begin(v), itv), 1u);
+        EXPECT_EQ(std::distance(std::begin(v), itv), 1);
         if (found) {
             EXPECT_EQ(*itv, 10);
         }
@@ -726,7 +726,7 @@ TEST(algorithms, binary_find)
         auto ita = binary_find(a, 10);
         auto found = ita!=std::end(a);
         EXPECT_TRUE(found);
-        EXPECT_EQ(std::distance(std::begin(a), ita), 1u);
+        EXPECT_EQ(std::distance(std::begin(a), ita), 1);
         if (found) {
             EXPECT_EQ(*ita, 10);
         }
@@ -735,7 +735,7 @@ TEST(algorithms, binary_find)
         auto itv = binary_find(v, 10);
         found = itv!=std::end(v);
         EXPECT_TRUE(found);
-        EXPECT_EQ(std::distance(std::begin(v), itv), 1u);
+        EXPECT_EQ(std::distance(std::begin(v), itv), 1);
         if (found) {
             EXPECT_EQ(*itv, 10);
         }
@@ -750,7 +750,7 @@ TEST(algorithms, binary_find)
         auto itv = binary_find(vr, 10);
         auto found = itv!=std::end(vr);
         EXPECT_TRUE(found);
-        EXPECT_EQ(std::distance(std::cbegin(v), itv), 1u);
+        EXPECT_EQ(std::distance(std::cbegin(v), itv), 1);
         if (found) {
             EXPECT_EQ(*itv, 10);
         }

--- a/test/unit/test_simd.cpp
+++ b/test/unit/test_simd.cpp
@@ -664,7 +664,7 @@ TYPED_TEST_P(simd_fp_value, fp_maths) {
 
         fp exprelr_u[N];
         for (unsigned i = 0; i<N; ++i) {
-            exprelr_u[i] = u[i]+fp(1)==fp(1)? fp(1): u[i]/(std::exp(u[i])-fp(1));
+            exprelr_u[i] = u[i]+fp(1)==fp(1)? fp(1): u[i]/(std::expm1(u[i]));
         }
         exprelr(simd(u)).copy_to(r);
         EXPECT_TRUE(testing::seq_almost_eq<fp>(exprelr_u, r));

